platform/entanglement/src/actions/CopyAction.js

```diff
@@ -34,16 +34,102 @@ define(
          * @constructor
          * @memberof platform/entanglement
          */
-        function CopyAction(locationService, copyService, context) {
-            return new AbstractComposeAction(
-                locationService,
-                copyService,
-                context,
-                "Duplicate",
-                "to a location"
-            );
+        function CopyAction($log, locationService, copyService, dialogService, 
+                            notificationService, context) {
+            this.dialog = undefined;
+            this.notification = undefined;
+            this.dialogService = dialogService;
+            this.notificationService = notificationService;
+            this.$log = $log;
+            //Extend the behaviour of the Abstract Compose Action
+            AbstractComposeAction.call(this, locationService, copyService, 
+                context, "Duplicate", "to a location");
         }
 
+        /**
+         * Updates user about progress of copy. Should not be invoked by
+         * client code under any circumstances.
+         *
+         * @private
+         * @param phase
+         * @param totalObjects
+         * @param processed
+         */
+        CopyAction.prototype.progress = function(phase, totalObjects, processed){
+            /*
+             Copy has two distinct phases. In the first phase a copy plan is
+             made in memory. During this phase of execution, the user is
+             shown a blocking 'modal' dialog.
+
+             In the second phase, the copying is taking place, and the user
+             is shown non-invasive banner notifications at the bottom of the screen.
+             */
+            if (phase.toLowerCase() === 'preparing' && !this.dialog){
+                this.dialog = this.dialogService.showBlockingMessage({
+                    title: "Preparing to copy objects",
+                    unknownProgress: true,
+                    severity: "info"
+                });
+            } else if (phase.toLowerCase() === "copying") {
+                this.dialogService.dismiss();
```

Comment:
Hm, this makes me want to rethink the dialogService API a little bit. Couple of things:

* `dismiss` is a private method of `dialogService`, so we shouldn't be relying on it here unless we make it public.
* The fact that `dismiss` is not connected to the specific dialog we previously displayed seems like it creates the opportunity for unusual behavior. I'm picturing some other blocking dialog being implemented in the same way (show the dialog on one event, dismiss it on another); if this were triggered at the same time as this dialog, then it would be very easy to dismiss the wrong dialog.

My thinking is that `showBlockingMessage` should return something that allows the specific dialog that was just shown to be dismissed. Either an object with a `dismiss` method, or perhaps just [a promise that we could pass into a `dialogService.cancel(promise)` function](http://www.bennadel.com/blog/2731-canceling-a-promise-in-angularjs.htm) (this would allow the other `dialogService` methods to also be cancelled consistently)
  
This is a little bit out of scope for this issue, so we could defer this into a follow-up issue, but since this is the first release usage of notifications we could also justify refactoring here. I'm comfortable with either option